osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,271 @@
+  if (!times.empty()) {
+    auto datetime = *times.begin();
+    auto time_vec = osquery::split(datetime, ";");
+
+    if (time_vec.size() == 1) {
+      auto _start = time_vec.front();
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start + "']");
+    } else if (time_vec.size() == 2) {
+      auto _start = time_vec.front();
+      auto _end = time_vec.at(1);
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start +
+                               "' and @SystemTime&lt;='" + _end + "']");
+    }
+  } else if (!timestamps.empty()) {
+    auto time_diff = *timestamps.begin();
+    xfilterList.emplace_back(
+        "TimeCreated[timediff(@SystemTime) &lt;= " + time_diff + "]");
+  }
+
+  xfilter = xfilterList.empty()
+                ? "*"
+                : "*[System[" + osquery::join(xfilterList, " and ") + "]]";
+}
+
+bool shouldHandleXpath(QueryContext& context) {
+  // xpaths are mutually execlusive and can't be used with itself
+  // and the other constraints
+  auto xpaths = context.constraints["xpath"].getAll(EQUALS);
+  if (xpaths.size() > 1) {
+    return false;
+  }
+
+  return !(context.hasConstraint("channel", EQUALS) ||
+           context.hasConstraint("time_range", EQUALS) ||
+           context.hasConstraint("timestamp", EQUALS));
+}
+
+void genWindowsEventLog(RowYield& yield, QueryContext& context) {
+  std::set<std::pair<std::string, std::string>> xpath_set;
+  auto hasXpath = context.hasConstraint("xpath", EQUALS);
+
+  if (hasXpath && !shouldHandleXpath(context)) {
+    LOG(WARNING) << "Error : xpaths are mutually exclusive and"
+                    " can't be used with other constraints "
+                    "(channel, time_range, timestamp)";
+    return;
+  }
+
+  // Check if the `xpath` constraint is available and query
+  // the events with xpath
+  if (hasXpath) {
+    auto xpaths = context.constraints["xpath"].getAll(EQUALS);
+    auto xpath = *xpaths.begin();
+    pt::ptree propTree;
+    std::stringstream ss;
+    ss << xpath;
+    pt::read_xml(ss, propTree);
+    auto channel = propTree.get("QueryList.Query.Select.<xmlattr>.Path", "");
+    if (!channel.empty()) {
+      xpath_set.insert(std::make_pair(channel, xpath));
+    } else {
+      LOG(WARNING) << "Invalid xpath format - " << xpath;
+    }
+
+  } else if (context.hasConstraint("channel", EQUALS)) {
+    auto channels = context.constraints["channel"].getAll(EQUALS);
+    std::string xfilter("");
+    genXfilterFromConstraints(context, xfilter);
+    std::string welSearchQuery = kEventLogXmlPrefix;
+
+    for (const auto& channel : channels) {
+      welSearchQuery += "<Select Path=\"" + channel + "\">";
+      welSearchQuery += xfilter;
+      welSearchQuery += "</Select>" + kEventLogXmlSuffix;
+      xpath_set.insert(std::make_pair(channel, welSearchQuery));
+    }
+
+  } else {
+    LOG(WARNING) << "must specify the event channel or xpath for lookup!";
```

Comment:
Should this be something like `"Query constraints are invalid: the event channel or xpath must be specified"`? Also prefer not to include punctuation in status messages, we assume they are all statements.